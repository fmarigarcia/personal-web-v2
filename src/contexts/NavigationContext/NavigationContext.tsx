import React, { createContext, useContext, useState } from 'react';
import type { ReactNode } from 'react';
import { SECTIONS } from '@utils/constants';

// Types
interface NavigationContextType {
    currentSection: string;
    setCurrentSection: (section: string) => void;
    isScrolling: boolean;
    setIsScrolling: (scrolling: boolean) => void;
}

// Create context
const NavigationContext = createContext<NavigationContextType | undefined>(
    undefined
);

// Provider component
interface NavigationProviderProps {
    children: ReactNode;
}

export const NavigationProvider: React.FC<NavigationProviderProps> = ({
    children,
}) => {
    const [currentSection, setCurrentSection] = useState<string>(SECTIONS.HERO);
    const [isScrolling, setIsScrolling] = useState<boolean>(false);

    const value: NavigationContextType = {
        currentSection,
        setCurrentSection,
        isScrolling,
        setIsScrolling,
    };

    return (
        <NavigationContext.Provider value={value}>
            {children}
        </NavigationContext.Provider>
    );
};

// Hook to use the context
export const useNavigation = (): NavigationContextType => {
    const context = useContext(NavigationContext);

    if (context === undefined) {
        throw new Error(
            'useNavigation must be used within a NavigationProvider'
        );
    }

    return context;
};
