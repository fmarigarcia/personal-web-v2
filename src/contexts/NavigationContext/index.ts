export {
    NavigationContext,
    type NavigationContextType,
} from './NavigationContext';
export { NavigationProvider } from './NavigationProvider';
export { useNavigation } from './useNavigation';
