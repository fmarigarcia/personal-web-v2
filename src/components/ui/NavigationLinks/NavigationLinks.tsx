import React from 'react';
import { useTranslation } from 'react-i18next';
import clsx from 'clsx';
import { useNavigation } from '@contexts/NavigationContext';
import { NAVIGATION_ITEMS } from '@utils/constants';

interface NavigationLinksProps {
    onNavClick: (sectionId: string) => void;
    className?: string;
}

export const NavigationLinks: React.FC<NavigationLinksProps> = ({
    onNavClick,
    className = '',
}) => {
    const { t } = useTranslation();
    const { currentSection } = useNavigation();

    return (
        <div
            className={clsx('hidden md:flex items-center space-x-8', className)}
        >
            {NAVIGATION_ITEMS.map((item) => (
                <button
                    key={item.id}
                    onClick={() => onNavClick(item.id)}
                    className={clsx(
                        'text-sm font-medium transition-colors hover:text-stone-950 dark:hover:text-green-400',
                        {
                            'text-stone-950 dark:text-green-400':
                                currentSection === item.id,
                            'text-stone-700 dark:text-green-200':
                                currentSection !== item.id,
                        }
                    )}
                    type="button"
                    aria-label={`Go to ${t(item.labelKey)} section`}
                >
                    {t(item.labelKey)}
                </button>
            ))}
        </div>
    );
};

export default NavigationLinks;
