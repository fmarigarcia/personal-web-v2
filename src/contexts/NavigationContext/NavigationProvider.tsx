import React, { useState, useCallback, useRef } from 'react';
import type { ReactNode } from 'react';
import { SECTIONS } from '@utils/constants';
import { NavigationContext } from './NavigationContext';
import type { NavigationContextType } from './NavigationContext';

interface NavigationProviderProps {
    children: ReactNode;
}

export const NavigationProvider: React.FC<NavigationProviderProps> = ({
    children,
}) => {
    const [currentSection, setCurrentSection] = useState<string>(SECTIONS.HERO);
    const [isScrolling, setIsScrolling] = useState<boolean>(false);

    // Use a ref to store the navigation function to avoid infinite loops
    const navigateRef = useRef<((sectionId: string) => void) | undefined>(
        undefined
    );

    // Stable navigation function that calls the current ref value
    const navigateToSection = useCallback((sectionId: string) => {
        if (navigateRef.current) {
            navigateRef.current(sectionId);
        }
    }, []);

    // Stable setter function that updates the ref
    const setNavigateToSection = useCallback(
        (fn: (sectionId: string) => void) => {
            navigateRef.current = fn;
        },
        []
    );

    const value: NavigationContextType = {
        currentSection,
        setCurrentSection,
        isScrolling,
        setIsScrolling,
        navigateToSection,
        setNavigateToSection,
    };

    return (
        <NavigationContext.Provider value={value}>
            {children}
        </NavigationContext.Provider>
    );
};
