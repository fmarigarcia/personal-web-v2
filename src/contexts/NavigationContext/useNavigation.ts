import { useContext } from 'react';
import { NavigationContext } from './NavigationContext';
import type { NavigationContextType } from './NavigationContext';

// Hook to use the context
export const useNavigation = (): NavigationContextType => {
    const context = useContext(NavigationContext);

    if (context === undefined) {
        throw new Error(
            'useNavigation must be used within a NavigationProvider'
        );
    }

    return context;
};
