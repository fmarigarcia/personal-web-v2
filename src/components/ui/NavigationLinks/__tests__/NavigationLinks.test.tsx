import { render, screen, fireEvent } from '@testing-library/react';
import { NavigationProvider } from '@contexts/NavigationContext';
import { NavigationLinks } from '../NavigationLinks';

// Mock i18next
jest.mock('react-i18next', () => ({
    useTranslation: () => ({
        t: (key: string) => {
            const translations: Record<string, string> = {
                'nav.home': 'Home',
                'nav.about': 'About',
                'nav.experience': 'Experience',
                'nav.contact': 'Contact',
            };
            return translations[key] || key;
        },
    }),
}));

describe('NavigationLinks', () => {
    const mockOnNavClick = jest.fn();

    beforeEach(() => {
        mockOnNavClick.mockClear();
    });

    const renderNavigationLinks = (props = {}) => {
        return render(
            <NavigationProvider>
                <NavigationLinks onNavClick={mockOnNavClick} {...props} />
            </NavigationProvider>
        );
    };

    it('should render all navigation items', () => {
        renderNavigationLinks();

        expect(screen.getByText('Home')).toBeInTheDocument();
        expect(screen.getByText('About')).toBeInTheDocument();
        expect(screen.getByText('Experience')).toBeInTheDocument();
        expect(screen.getByText('Contact')).toBeInTheDocument();
    });

    it('should handle navigation clicks', () => {
        renderNavigationLinks();

        const aboutButton = screen.getByText('About');
        fireEvent.click(aboutButton);

        expect(mockOnNavClick).toHaveBeenCalledWith('about');
    });

    it('should highlight current section', () => {
        renderNavigationLinks();

        // Hero should be highlighted by default
        const homeButton = screen.getByText('Home');
        expect(homeButton).toHaveClass('text-stone-950');

        // Other buttons should not be highlighted
        const aboutButton = screen.getByText('About');
        expect(aboutButton).toHaveClass('text-stone-700');
    });

    it('should have proper accessibility attributes', () => {
        renderNavigationLinks();

        const buttons = screen.getAllByRole('button');
        buttons.forEach((button) => {
            expect(button).toHaveAttribute('type', 'button');
            expect(button).toHaveAttribute('aria-label');
        });
    });

    it('should apply custom className', () => {
        const customClass = 'custom-nav';
        const { container } = renderNavigationLinks({ className: customClass });

        const navContainer = container.querySelector('div');
        expect(navContainer).toHaveClass(customClass);
    });

    it('should be hidden on mobile by default', () => {
        const { container } = renderNavigationLinks();

        const navContainer = container.querySelector('div');
        expect(navContainer).toHaveClass('hidden', 'md:flex');
    });

    it('should have hover styles', () => {
        renderNavigationLinks();

        const buttons = screen.getAllByRole('button');
        buttons.forEach((button) => {
            expect(button).toHaveClass(
                'hover:text-stone-950',
                'transition-colors'
            );
        });
    });

    it('should match snapshot', () => {
        const { container } = renderNavigationLinks();
        expect(container.firstChild).toMatchSnapshot();
    });
});
