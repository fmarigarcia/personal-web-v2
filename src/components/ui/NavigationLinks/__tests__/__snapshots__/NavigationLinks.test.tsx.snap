// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationLinks should match snapshot 1`] = `
<div
  class="hidden md:flex items-center space-x-8"
>
  <button
    aria-label="Go to Home section"
    class="text-sm font-medium transition-colors hover:text-blue-600 text-blue-600"
    type="button"
  >
    Home
  </button>
  <button
    aria-label="Go to About section"
    class="text-sm font-medium transition-colors hover:text-blue-600 text-gray-700"
    type="button"
  >
    About
  </button>
  <button
    aria-label="Go to Experience section"
    class="text-sm font-medium transition-colors hover:text-blue-600 text-gray-700"
    type="button"
  >
    Experience
  </button>
  <button
    aria-label="Go to Contact section"
    class="text-sm font-medium transition-colors hover:text-blue-600 text-gray-700"
    type="button"
  >
    Contact
  </button>
</div>
`;
